### PR TITLE
WT-7190 Limit eviction of non-history store pages when checkpoint is operating on history store

### PR DESCRIPTION
**Pull request by hbokhari**
If we are trying to evict a dirty page that does not belong to history store(HS) and checkpoint is processing HS file, then avoid evicting the dirty non-HS page for now if the cache is already dominated by dirty HS content. Evicting a non-HS dirty page can generate even more HS content. As we can not evict HS pages while checkpoint is operating on the HS file, we can end up in a situation where we exceed the cache size limits.

### CODE REVIEW

**Comment by hbokhari** [comment]
@quchenhao @kommiharibabu @bvpvamsikrishna - This is my go at finding a solution for WT-7190. If we can agree on the solution, I'll start MongoDB patch builds to verify it further.

File: src/evict/evict_page.c [hunk 1]
```diff
@@ -575,6 +575,18 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
...
+    /*
+     * If we trying to evict a dirty page that does not belong to history store(HS) and checkpoint
+     * is processing HS file, then avoid evicting the dirty non-HS page for now if the cache is
+     * already dominated by dirty HS content.
+     *
+     * Evicting a non-HS dirty page can generate even more HS content. As we can not evict HS pages
+     * while checkpoint is operating on the HS file, we can end up in a situation where we exceed
+     * the cache size limits.
+     */
+    if (modified && !WT_IS_HS(btree->dhandle) && __wt_cache_hs_dirty(session) &&
```
Comment (quchenhao) [inline src/evict/evict_page.c:587]: How do you know the cache is dominated by hs content? This change I think effectively blocks evicting data store tables if checkpoint is working on the hs table. If it takes a long time to checkpoint the hs table, then we may run into cache pressure.

Comment (sulabhM) [reply src/evict/evict_page.c:587]: `__wt_cache_hs_dirty(session)` returns true if HS content in cache > 20%

**Comment by sulabhM** [comment]
I have a feeling this change will lead to the stall of the workload when HS content crosses 20%, for the time checkpoint is running on the HS. I also agree with @quchenhao, checkpoints on HS might take significant time in the real world. All application threads will spin repeatedly trying to evict non-HS pages, and fail to do so. I could be wrong, needs to be tested out.

Having said that, I have only read the WT ticket just now, and have not put thought to any alternative solutions.

File: src/evict/evict_page.c (hunk 1 above)
Comment (quchenhao) [reply src/evict/evict_page.c:587]: OK, got it. Does 20% is high for hs?

Comment (hbokhari) [reply src/evict/evict_page.c:587]: Note that `__wt_cache_hs_dirty()` is returning true when the dirty content from HS alone is 20% or more. We generally want to keep overall WT dirty content below 20% to keep the memory fragmentation to minimum.

**Comment by hbokhari** [comment]
I agree that this change can potentially stall every other thread except for the checkpoint thread. We can probably make the check narrower to include overall cache usage as well (e.g., overall cache occupancy >95%) .

If we continue evicting pages in this situation, our cache usage jumps beyond 100%. I have observed cache usage jumping to 185%. So we have to balance between not able to evict the page and possibly stall, or let the cache usage run away.

**Comment by quchenhao** [comment]
> I agree that this change can potentially stall every other thread except for the checkpoint thread. We can probably make the check narrower to include overall cache usage as well (e.g., overall cache occupancy >95%) .
> 
> If we continue evicting pages in this situation, our cache usage jumps beyond 100%. I have observed cache usage jumping to 185%. So we have to balance between not being able to evict the page and possibly stall, or letting the cache usage run away.

I agree adding the overall cache usage check makes sense to me. However, I'm still concerned in that case, we will stall as we can't do any eviction until we evict the history store pages. I think we already prioritising evicting history store pages, don't we? @sulabhM If we haven't, we can think of prioritising that.

**Comment by hbokhari** [comment]
> I think we already prioritising evicting history store pages, don't we? @sulabhM If we haven't, we can think of prioritising that.

That is correct - we prfioritize evicting HS pages when HS start dominating the cache (WT-6924). However, that mechanism would not work in this scenario because we can not evict pages from a tree that is being processed by the checkpoint.

**Comment by sulabhM** [comment]
We do already prioritize HS pages for eviction when HS content is more than 20%.

If we check for an overall 95%, we will completely stall the workload at 95%. As of current behavior, we throttle at 95% using application threads and completely stall at 100%.

```
         /* See if eviction is still needed. */                                                       
         if (!__wt_eviction_needed(session, busy, readonly, &pct_full) ||                             
           (pct_full < 100.0 && (cache->eviction_progress > initial_progress + max_progress)))        
             break;    
```

The percentage full is also calculated a bit different, in `__wt_eviction_needed`:

```
     /*                                                                                               
      * Calculate the cache full percentage; anything over the trigger means we involve the           
      * application thread.                                                                           
      */                                                                                              
     if (pct_fullp != NULL)                                                                           
         *pct_fullp = WT_MAX(0.0,                                                                     
           100.0 -                                                                                    
             WT_MIN(                                                                                  
               WT_MIN(cache->eviction_trigger - pct_full, cache->eviction_dirty_trigger - pct_dirty), 
               cache->eviction_updates_trigger - pct_updates));
```

The read workload will also entirely stall, unfortunately.

Maybe we can try checking against 95% and 100% and at least know if comparing against those would resolve the issue.

**Comment by hbokhari** [comment]
@sulabhM - Thanks for the comment. I have added a check for overall cache using `__wt_eviction_needed()` and I can see we are just touching the limit of cache size.
image

My only concern is that if we check at let say close to 100% usage, we might already have crossed the size limit. But I'll run a test with my reproducer to check.

**Comment by hbokhari** [comment]
@sulabhM, this is what cache usage looks like when I included `__wt_cache_full()` in the check. The cache usage stays at 100.75% which in my opinion is still acceptable. Essentially we will only block any further operations on the database in this scenario if the cache is fully utilised.

image

**Comment by quchenhao** [comment]
I'm worried that if oldest timestamp is not moved, i.e., we have long point in time reads or secondary is down, checkpointing hs will take much longer and we stall other operations frequently.

**Comment by hbokhari** [comment]
Based on the reproducer script, the cache size violation happens for almost ~70% of all checkpoints. Therefore, I don't think aborting checkpoint will be a valid option here. 

Note that the script I am using for testing is written is to heavily exercise HS and might be an extreme case.

**Comment by sulabhM** [comment]
This is a good discussion, and we have investigated a possible solution. We usually have design discussions in the Jira ticket. Maybe we should move there for further discussions and lock the PR till we agree on a path forward. @hbokhari, if you agree with me, can we summarize these experiments and findings on the jira and continue there.

**Comment by hbokhari** [comment]
@sulabhM I agree we need move the discussion to JIRA. I have summarised the findings there.

File: src/txn/txn_ckpt.c [hunk 2]
```diff
@@ -897,7 +897,13 @@ __txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
      */
     if (F_ISSET(hs_dhandle, WT_DHANDLE_OPEN)) {
         time_start_hs = __wt_clock(session);
+        conn->txn_global.checkpoint_running_hs = true;
```
Comment (quchenhao) [inline src/txn/txn_ckpt.c:900]: Do you need a memory barrier here?

File: src/evict/evict_page.c [hunk 3]
```diff
@@ -575,6 +575,20 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
...
+     * is processing HS file, then avoid evicting the dirty non-HS page for now if the cache is
+     * already dominated by dirty HS content.
+     *
+     * Evicting a non-HS dirty page can generate even more HS content. As we can not evict HS pages
+     * while checkpoint is operating on the HS file, we can end up in a situation where we exceed
+     * the cache size limits.
+     */
+    if (modified && conn->txn_global.checkpoint_running_hs && !WT_IS_HS(btree->dhandle) &&
+      __wt_cache_hs_dirty(session) && __wt_cache_full(session)) {
+        WT_STAT_CONN_INCR(session, cache_eviction_checkpoint_hs);
```
Comment (sulabhM) [inline src/evict/evict_page.c:589]: Rename cache_eviction_checkpoint_hs to cache_eviction_blocked_checkpoint_hs OR some other name that has word "blocked" in it.

**Review by sulabhM — APPROVED** [review]

File: src/evict/evict_page.c (hunk 3 above)
Comment (hbokhari) [reply src/evict/evict_page.c]: Fixed.

File: src/txn/txn_ckpt.c (hunk 2 above)
Comment (hbokhari) [reply src/txn/txn_ckpt.c:900]: Interesting question - by the looks of it we can guarantee that `conn->txn_global.checkpoint_running_hs` will be set to true before we call `__wt_checkpoint()`. However, [we are not using memory barrier](https://github.com/wiredtiger/wiredtiger/blob/ba88c91101ff0ee6e24373e701b1f9537cc46f3b/src/txn/txn_ckpt.c#L1101) in `__txn_checkpoint_wrapper()` as well where we are setting/clearing `txn_global->checkpoint_running` flag.

Comment (sulabhM) [reply src/txn/txn_ckpt.c:900]: I do not think we need one. There are no precise timing considerations for functional correctness here. It is not that we are waiting for the exact spot in the code where we disable eviction on this file. Worse we would let a non-HS page be evicted and add some more to the HS content in the cache that can't evict.

Comment (quchenhao) [reply src/txn/txn_ckpt.c:900]: That's true. Good point.

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by hbokhari** [comment]
@bvpvamsikrishna can you have a final look?

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**